### PR TITLE
docs: sync all documentation with final implementation

### PR DESCRIPTION
**Pull request by melroyanthony**
## Summary

Comprehensive documentation sync — fixes 10 categories of staleness across 14 files.

### Fixes
1. Test count 15 → 18 everywhere
2. ADR-004: K_cap=min superseded by K_max=max (ADR-006)
3. data-flow.md: K_cap → K_max
4. README benchmark results: current latencies (262K, 909K, 12.4M, 4.7M, 815K)
5. CHANGELOG: test count, 3 new ADRs, 3 new tests, latencies
6. workspace.dsl: granularity "(w,h)" → "(w,h,k)"
7. security-model.md: Track B Gemini API calls
8. ADR-003: Python → Rust/Python, cost-based note
9. system-design.md: removed draft text, bench 17 ops 96→103
10. database-schema.md: bench 13 tensors 96→100

All 18 Rust tests passing. No code changes.

### CODE REVIEW

File: solution/docs/decisions/ADR-003-greedy-fusion.md [hunk 1]
```diff
@@ -41,7 +41,7 @@ A merge is only valid if the resulting subgraph is a **connected directed subDAG
 ## Consequences
 
 ### Positive
-- **Simple to implement**: ~150 lines of Python, well within the time budget
+- **Simple to implement**: implemented in Rust (Track A) and Python (Track B), well within the time budget
 - **Fast to execute**: O(N^2) worst case, sub-second for N=96
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/docs/decisions/ADR-003-greedy-fusion.md]: Fixed: N=96 → N=103.

File: solution/docs/decisions/ADR-003-greedy-fusion.md [hunk 2]
```diff
@@ -59,3 +59,20 @@ A merge is only valid if the resulting subgraph is a **connected directed subDAG
...
+
+## Enhancement: Cost-Based Fusion with Epsilon Tolerance
+
+Since the initial ADR was written, the merge criterion was strengthened (Issue #16).
+The pure feasibility check (merge if working set fits) was replaced with a
+**cost-based merge criterion**: merge subgraphs A and B only when
+`latency(A+B, best_gran_fused) < latency(A, best_gran_A) + latency(B, best_gran_B)`.
+
+An epsilon tolerance is applied so that merges with negligible latency difference
+(within floating-point rounding) are accepted, avoiding churn on borderline cases.
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/docs/decisions/ADR-003-greedy-fusion.md]: Fixed: wording now says merge requires lat_fused < lat_split - eps (rejects borderline cases, not accepts them).

File: solution/docs/architecture/security-model.md [hunk 3]
```diff
@@ -1,6 +1,15 @@
...
+This is a single-user CLI tool that processes local JSON files. Track A (Rust
+binary) has no network communication, no authentication, no multi-tenancy, and
+no user-supplied code execution.
+
+Track B (Python agent) makes HTTPS calls to the Gemini API
+(`generativelanguage.googleapis.com`) when a `GOOGLE_API_KEY` environment
+variable is set. No problem data or solution JSON is logged or persisted by the
+API client beyond the scope of a single agent run. When `GOOGLE_API_KEY` is set
+to `dummy` or omitted, the agent runs in local-only mode with zero network
+traffic.
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/docs/architecture/security-model.md]: Fixed: added note about error-path response preview (first 500 chars to stderr). Clarified as 'not routine logging'.

File: solution/checkpoints/stage-4-validation.md [hunk 4]
```diff
@@ -71,6 +72,9 @@
 | test_edge_fusion_ephemeral_correctness | Tensor 3 ephemeral in fused [0,1] |
 | test_edge_cyclic_dag_rejected | Cyclic input returns Err("cycle") |
 | test_benchmark_solutions_validity | All 5 benchmarks: full op coverage, valid JSON |
+| test_fused_matmul_pointwise_splitk | Fused MatMul+Pointwise with split-K granularity |
+| test_boundary_pw_input | Boundary Pointwise input tensor memory accounting |
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/checkpoints/stage-4-validation.md]: Fixed: test name corrected to test_fused_matmul_pointwise_splitk_boundary_pw_input.

File: solution/CHANGELOG.md [hunk 5]
```diff
@@ -82,6 +82,8 @@ Format follows [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
     - Edge cases: single tiny op, OOM detection, serialization round-trip,
       ephemeral tensor boundary correctness, cyclic DAG rejection
     - All 5 released benchmarks: full pipeline validity check
+    - New tests (3): `fused_matmul_pointwise_splitk`, `boundary_pw_input`,
+      `mixed_k_two_matmuls`
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/CHANGELOG.md]: Fixed: all 3 test names now use full function names with test_ prefix.

File: solution/docs/decisions/ADR-004-k-dimension-search.md [hunk 6]
```diff
@@ -23,14 +28,14 @@ candidates = w_values x h_values x k_values
 Where:
 - `w_values`: powers of 2 up to output width (as before)
 - `h_values`: powers of 2 up to output height (as before)
-- `k_values`: K_cap, K_cap/2, K_cap/4, ..., 1 where K_cap = min(K_full for each MatMul in the subgraph)
+- `k_values`: K_max, K_max/2, K_max/4, ..., 1 where K_max = max(K_full for each MatMul in the subgraph)
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/docs/decisions/ADR-004-k-dimension-search.md]: Fixed: k_values described as 'repeated integer halving from K_max' instead of 'powers of 2'.

File: solution/docs/decisions/ADR-004-k-dimension-search.md [hunk 7]
```diff
@@ -23,14 +28,14 @@ candidates = w_values x h_values x k_values
...
 
 For each (w, h, k) candidate:
 1. Check the working set fits in fast memory (OOM constraint)
 2. Compute total subgraph latency as the sum of per-step roofline costs across all tiles and k-steps
 3. Select the (w, h, k) triple that minimizes total subgraph latency
 
-Using min(K_full) across MatMuls ensures k never exceeds any op's reduction dimension. Note: k candidates are powers of 2, so `ceil(K_full / k)` correctly handles cases where k does not evenly divide K_full — the last k-step simply processes the remainder.
+Using max(K_full) across MatMuls as the upper bound (K_max) drives the subgraph for the full extent of the largest reduction dimension. MatMuls with smaller K_full values become inactive once they finish their k-steps, contributing zero compute and memory on subsequent steps. See ADR-006 for the mixed-K execution model. Note: k candidates are powers of 2, so `ceil(K_full / k)` correctly handles cases where k does not evenly divide K_full — the last k-step simply processes the remainder.
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/docs/decisions/ADR-004-k-dimension-search.md]: Fixed: removed 'powers of 2' claim. Now says 'generated by repeated integer halving (not necessarily powers of 2 unless K_max is)'.

File: solution/docs/architecture/data-flow.md [hunk 8]
```diff
@@ -63,7 +63,7 @@ sequenceDiagram
     S-->>G: ScheduleState (with split-K)
 
     G->>G: For each subgraph, generate (w, h, k) candidates
-    Note over G: k candidates: K_cap down to 1 in powers of 2<br/>K_cap = min(K_full across MatMuls in subgraph)
+    Note over G: k candidates: K_max down to 1 in powers of 2<br/>K_max = max(K_full across MatMuls in subgraph)
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/docs/architecture/data-flow.md]: Fixed: diagram text changed to 'K_max down to 1 by repeated halving'. Flowchart label also updated.

File: solution/docs/architecture/security-model.md [hunk 9]
```diff
@@ -1,6 +1,17 @@
...
+no user-supplied code execution.
+
+Track B (Python agent) makes HTTPS calls to the Gemini API
+(`generativelanguage.googleapis.com`) when a `GOOGLE_API_KEY` environment
+variable is set. No problem data or solution JSON is logged or persisted by the
+API client beyond the scope of a single agent run. When `GOOGLE_API_KEY` is set
+to `dummy` or omitted, the agent runs in local-only mode with zero network
+traffic. Note: on API error paths, the agent may print a preview of the
+Gemini response (first 500 chars) to stderr for debugging, which could include
+partial solution JSON. This is not routine logging.
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/docs/architecture/security-model.md]: Fixed: clarified Track B transmits full problem JSON and solution JSON to Gemini when API key is set. Removed 'not logged/persisted' implication.

File: solution/README.md [hunk 10]
```diff
@@ -336,18 +336,18 @@ Validation checks per output file:
...
-for MatMul data reuse.
+| Benchmark | Ops | Tensors | Fast Mem | Bandwidth | Track A Latency | Subgraphs |
+|-----------|-----|---------|----------|-----------|-----------------|-----------|
+| mlsys-2026-1  | 5   | 9   | 60,000   | 20  | 262,822    | 4  |
+| mlsys-2026-5  | 19  | 29  | 30,000   | 15  | 909,261    | 13 |
+| mlsys-2026-9  | 32  | 49  | 250,000  | 25  | 12,415,140 | 24 |
+| mlsys-2026-13 | 63  | 100 | 600,000  | 50  | 4,707,779  | 25 |
+| mlsys-2026-17 | 103 | 160 | 500,000  | 100 | 814,572    | 81 |
+
+All benchmarks complete in under 1 second. The optimizer uses cost-based
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/README.md]: Fixed: changed to 'The optimizer runs in under 1 second per benchmark on a standard dev machine'.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -52,7 +52,7 @@ jobs:
         working-directory: solution/backend/rust
         run: cargo build --release
 
-      - name: Run unit tests (15 tests)
+      - name: Run unit tests (18 tests)
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply .github/workflows/ci.yml]: Fixed: removed hardcoded count from step name. Now just 'Run unit tests'.

File: solution/docs/decisions/ADR-004-k-dimension-search.md [hunk 12]
```diff
@@ -14,7 +19,7 @@ Benchmark analysis showed:
 The root cause was that the search evaluated candidates but k=1 minimizes the per-step working set (smallest slices). The search was not properly accounting for the multiplicative k-step count and the repeated strip reloading that comes with smaller k values.
 
 ## Decision
-Search k from min(K_full across all MatMuls in the subgraph) down to 1 in powers of 2, jointly with (w, h) spatial candidates. The full search space becomes:
+Search k from max(K_full across all MatMuls in the subgraph) down to 1 in powers of 2, jointly with (w, h) spatial candidates. The full search space becomes:
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/docs/decisions/ADR-004-k-dimension-search.md]: Fixed: Decision sentence now says 'by repeated integer halving'.
